src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,103 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
+    public void checkCapacity() {
+        if (elementData.length <= size) {
+            int newCapacity = (int) (elementData.length * GROWTH_FACTOR);
+            elementData = Arrays.copyOf(elementData, newCapacity);
+        }
+    }
+
+    public void checkIndex(int index) {
```

Comment:
The `checkIndex` method is good for checking the index for get, set, or remove operations, but for adding an element at a specific index, the valid range should be `0` to `size` inclusive. The current check will throw an exception if trying to add at the end of the list (index == size), which should be allowed.